### PR TITLE
Device CRD: Add types for devices

### PR DESCRIPTION
**Pull request by rohitsardesai83**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
Add device crd apis and yamls.

**Which issue(s) this PR fixes**:
Address #357 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by sids-b** [comment]
This PR will get modified based on #362, so we need to focus on 362.

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/register.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen=package,register
+// +k8s:conversion-gen=kubeedge/cloud/edgecontroller/pkg/devicecontroller/apis/devices
+// +k8s:defaulter-gen=TypeMeta
+// +groupName=devices.kubeedge.io
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	//"sigs.k8s.io/controller-runtime/pkg/runtime/scheme"
```
Comment (m1093782566) [inline cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/register.go]: Please remove the deadline.

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+// DeviceModelList contains a list of DeviceModel
+type DeviceModelList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DeviceModel `json:"items"`
+}
+
+func init() {
+	//SchemeBuilder.Register(&Device{}, &DeviceList{}, &DeviceModel{})
```
Comment (m1093782566) [inline cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go]: We are not registering device and devicemodel here?

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	// Required.
+	IP string `json:"ip,omitempty"`
+	// Required.
+	Port int64 `json:"port,omitempty"`
+	// Required.
+	SlaveID string `json:"slaveID,omitempty"`
+}
+
+type ProtocolConfigModbusRTU struct {
+
```
Comment (m1093782566) [inline cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go]: nit: please remove the empty line.

**Comment by m1093782566** [comment]
For the crd sample yamls in the `/build` folder, we are not using them right? What's the usage of them?

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+// DeviceStatus reports the device state and the desired/reported values of twin attributes.
+type DeviceStatus struct {
+	// A list of device twins containing desired/reported desired/reported values of twin properties..
+	// Optional: A passive device won't have twin properties and this list could be empty.
+	// +optional
+	Twins []Twin `json:"twins,omitempty"`
+}
+
+// A Twin provides a logical representation of control properties (writable properties in the
```
Comment (m1093782566) [inline cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go]: Remove `A` to get golint check passed in the future we add static checks in CI.

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go (hunk 2 above)
Comment (sids-b) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go]: no, these generated files require k8s version 1.11 or above, so we will register them in the controller code.

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/register.go (hunk 1 above)
Comment (rohitsardesai83) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/register.go]: done

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go (hunk 2 above)
Comment (rohitsardesai83) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go]: this code depends on   "sigs.k8s.io/controller-runtime" project which has k8s dependency >= 1.11.
our current k8s dependency is 1.10.9, so we can't use the controller-runtime package for scheme related operations. we are doing this manually as Sid mentioned

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go (hunk 4 above)
Comment (rohitsardesai83) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go]: updated

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go (hunk 2 above)
Comment (m1093782566) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go]: So we can remove the `init()` here?

File: cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go (hunk 3 above)
Comment (rohitsardesai83) [reply cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_instance_types.go]: fixed

**Comment by rohitsardesai83** [comment]
> For the crd sample yamls in the `/build` folder, we are not using them right? What's the usage of them?

these can be used for our tests to create devices and models.

**Comment by m1093782566** [comment]
@rohitsardesai83 

Please squash the commits.

**Comment by kevin-wangzefeng** [comment]
Accepted with squash-merge :-)

**Pull request merged**